### PR TITLE
Update the SolidusPaypalBraintree namespace to SolidusBraintree

### LINKED ISSUE #96: Update the SolidusPaypalBraintree namespace to SolidusBraintree
## Goal

We want to rename the SolidusPaypalBraintree namespace to SolidusBraintree to simplify the extension's name and avoid it being confused with SolidusPaypalCommercePlatform.

## Requirements

* [x] For backward compatibility, we also need to contextually allow people to choose the namespace for database table names, so that those who are using the `solidus_paypal_braintree_prefix` can still update to the latest version. 
* [x] We want to limit the manual changes that users of SolidusPaypalBraintree have to make in order to transition to SolidusBraintree. We may want to consider to alias the SolidusPaypalBraintree class and the solidus_paypal_braintree JS file names. The aliases have to be deprecated.
* [x] We want to ensure that existing payments to SolidusPaypalBraintree will work with the new name. We may want to add a data migration to rename the source type of these payments from SolidusPaypalBraintree to SolidusBraintree.

## Dependencies

This is blocked by #92.

### PR DESCRIPTION
**Pull request by gsmendoza**
## Summary

Resolves #96.

## Goal

We want to rename the SolidusPaypalBraintree namespace to SolidusBraintree to simplify the extension's name and avoid it being confused with SolidusPaypalCommercePlatform.

## Requirements

* [x] For backward compatibility, we also need to contextually allow people to choose the namespace for database table names, so that those who are using the `solidus_paypal_braintree_prefix` can still update to the latest version. 
* [x] We want to limit the manual changes that users of SolidusPaypalBraintree have to make in order to transition to SolidusBraintree. We may want to consider to alias the SolidusPaypalBraintree class and the solidus_paypal_braintree JS file names. The aliases have to be deprecated.
* [x] We want to ensure that existing payments to SolidusPaypalBraintree will work with the new name. We may want to add a data migration to rename the source type of these payments from SolidusPaypalBraintree to SolidusBraintree.

## Dependencies

This is blocked by #92.

## Demo

See https://drive.google.com/file/d/1qWl09L7OSyMwH2G1t5jvjD1hu8QL7REv/view?usp=share_link.

Demo app code can be found at https://github.com/gsmendoza/solidus_store/tree/rails-7-0-4--solidus-3-2-5--braintree--renamed--minimal.

## Post-review TODOs

* [x] ~Try deprecating SCSS files using https://sass-lang.com/documentation/at-rules/warn.~ - Removing the assets files will be automated.
* [x] ~Rename the tables by adding more migrations.~ - Won't continue since renaming tables can be a potentially dangerous operation. It's best to leave this task to the extension's user.
* [x] Use InstallGenerator to update installation.
* [x] ~Use `attr_accessor` for `table_name_prefix`.~ - We'll no allow users to configure the table name prefix.
* [x] Update README: document what to do with the previous gem.
* [x] ~Advise that upgrading solidus_braintree from 1.2.0 to 2.0.0 is not recommended. Rather 2.0.0 is intended for new stores and migrations from solidus_paypal_braintree.~ - With the data migrations decoupled from the models, it should now be possible to migrate directly from any old version of the gem.
* [x] Include the target version requirement in the upgrade instructions.

## Post-review TODOs (2023-01-024)

Table name prefix:

* [x] Do not change the table name prefix in old migrations.
* [x] Remove the `table_name_prefix` configuration - we don't want to encourage users to change the table name prefix, because in the future we would still be adding migrations that use the old prefix.
* [x] Document that we have chosen to keep the table name prefix to `solidus_paypal_braintree` in order to minimize disruption during the migration to the new version.

SolidusPaypalBraintree references:

* [x] Ensure that, except for the table name prefix, we don't leave any references to SolidusPaypalBraintree in the user's app, except for those references the user has added himself.
* [x] Rename the `solidus_paypal_braintree` initializer to `solidus_braintree`.

Upgrade instructions announcement:

* [x] Add a post-install message to the gem linking to the README update section.

Testing:

* ~[ ] Test what would happen if the user adds both `solidus_braintree` and `solidus_paypal_braintree` to the app.~ - Skipping since this edge case is highly unlikely.

## Post-review TODOs (2023-01-31)

* [x] Update commit message to "Use rake and mount".
* [x] Add optional `SolidusPaypalBraintree` alias.
* [x] Comment out the table updates with a warning saying that they can lock the tables affected.

## Checklist

Check out our [PR guidelines](https://github.com/solidusio/.github/blob/master/CONTRIBUTING.md#pull-request-guidelines) for more details.

The following are mandatory for all PRs:

- [x] I have written a thorough PR description.
- [x] I have kept my commits small and atomic.
- [x] I have used clear, explanatory commit messages.

The following are not always needed (~cross them out~ if they are not):

- ~[ ] I have added automated tests to cover my changes.~
- [x] I have attached screenshots to demo visual changes.
- ~[ ] I have opened a PR to update the [guides](https://github.com/solidusio/edgeguides).~
- [x] I have updated the README to account for my changes.

### CODE REVIEW

File: db/migrate/20160830061749_create_solidus_paypal_braintree_sources.rb [hunk 1]
```diff
@@ -1,6 +1,6 @@
 class CreateSolidusPaypalBraintreeSources < SolidusSupport::Migration[4.2]
```
Comment (gsmendoza) [inline db/migrate/20160830061749_create_solidus_paypal_braintree_sources.rb:1]: I didn't rename the migration filenames and class names because if renamed, `bin/rails railties:install:migrations` would interpret the migrations as new migrations for SolidusBraintree.

File: app/assets/stylesheets/spree/backend/solidus_paypal_braintree.scss [hunk 2]
```diff
@@ -1,28 +1,4 @@
 /*
-Placeholder manifest file.
-the installer will append this file to the app vendored assets here: 'vendor/assets/stylesheets/spree/backend/all.css'
+ * This CSS file is deprecated. Require spree/backend/solidus_braintree instead.
```
Comment (gsmendoza) [inline app/assets/stylesheets/spree/backend/solidus_paypal_braintree.scss:2]: I don't know if we have a convention or tool for deprecating CSS files.

File: app/assets/javascripts/spree/backend/solidus_paypal_braintree.js [hunk 3]
```diff
@@ -1,96 +1,2 @@
...
-        });
-      } else {
-        // If we don't have previous cards, init fields immediately
-        initFields($this, id);
-        showForm(id);
-      }
-    });
-  });
-});
+// @deprecated since 2.0.0. Require spree/backend/solidus_braintree instead.
```
Comment (gsmendoza) [inline app/assets/javascripts/spree/backend/solidus_paypal_braintree.js:1]: Deprecating JS files using [JSDoc `@deprecated` tag](https://jsdoc.app/tags-deprecated.html).

File: db/migrate/20230109080950_rename_solidus_paypal_braintree_source_type.rb [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+  def up
+    Spree::Payment.where(source_type: 'SolidusPaypalBraintree::Source')
+      .update_all("source_type = 'SolidusBraintree::Source'")
+
+    Spree::WalletPaymentSource.where(payment_source_type: 'SolidusPaypalBraintree::Source')
+      .update_all("payment_source_type = 'SolidusBraintree::Source'")
+
+    Spree::PaymentMethod.where('type = ?', 'SolidusPaypalBraintree::Gateway')
+      .update_all("type = 'SolidusBraintree::Gateway'")
+  end
```
Comment (gsmendoza) [inline db/migrate/20230109080950_rename_solidus_paypal_braintree_source_type.rb:11]: These are the only tables I encountered during testing that point to the payment source type. Did I miss any more?

File: db/migrate/20230109080950_rename_solidus_paypal_braintree_source_type.rb [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+
+  def down
+    raise ActiveRecord::IrreversibleMigration
```
Comment (gsmendoza) [inline db/migrate/20230109080950_rename_solidus_paypal_braintree_source_type.rb:14]: Not allowing the data migration to be reverted to be safe: it's best not to assume what data previously existed in the user's database before the migration was ran.

File: lib/generators/solidus_braintree/update/update_generator.rb [hunk 6]
```diff
@@ -0,0 +1,23 @@
+# frozen_string_literal: true
+
+module SolidusBraintree
+  module Generators
+    class UpdateGenerator < Rails::Generators::Base
```
Comment (gsmendoza) [inline lib/generators/solidus_braintree/update/update_generator.rb:5]: Added an UpdateGenerator to add the data migration to the user's app.

File: README.md [hunk 7]
```diff
@@ -1,27 +1,44 @@
...
+1. Change the gem in your Gemfile from 'solidus_paypal_braintree' to 'solidus_braintree'.
+2. Run `bin/rails g solidus_braintree:update`. This will run a data migration updating `SolidusPaypalBraintree` references to `SolidusBraintree`.
+3. To support your existing database tables, set the `table_name_prefix` to `'solidus_paypal_braintree_'`:
+
+      ```rb
+      SolidusPaypalBraintree.configure do |config|
+        config.table_name_prefix = 'solidus_paypal_braintree_'
+      end
+      ```
+4. Update the engine route to `mount SolidusBraintree::Engine, at: '/solidus_braintree'`.
```
Comment (gsmendoza) [inline README.md:38]: I intentionally avoided automating this with the UpdateGenerator because I avoided allowing the UpdateGenerator to make changes to the user's app code. It can be hard to predict what changes the user has made to their app, and testing for those changes can make the UpdateGenerator more complicated than necessary. Up for discussion :)

File: app/assets/stylesheets/spree/backend/solidus_paypal_braintree.scss (hunk 2 above)
Comment (kennyadsl) [reply app/assets/stylesheets/spree/backend/solidus_paypal_braintree.scss:2]: We could explore using https://sass-lang.com/documentation/at-rules/warn somehow when that file is loaded.

File: lib/solidus_paypal_braintree.rb [hunk 8]
```diff
@@ -0,0 +1,4 @@
+# frozen_string_literal: true
+
+SolidusPaypalBraintree =
+  ActiveSupport::Deprecation::DeprecatedConstantProxy.new('SolidusPaypalBraintree', 'SolidusBraintree')
```
Comment (kennyadsl) [inline lib/solidus_paypal_braintree.rb:4]: ❤️

File: README.md (hunk 7 above)
Comment (kennyadsl) [reply README.md:38]: It's nice having these instructions here, but what about having them also in the installer? I mean printing them in the prompt when the update command is executed.

File: README.md [hunk 9]
```diff
@@ -12,7 +12,7 @@
 Add solidus_braintree to your Gemfile:
 
 ```ruby
-gem 'solidus_braintree', github: 'solidusio/solidus_braintree', branch: :master
+gem 'solidus_braintree'
```
Comment (kennyadsl) [inline README.md:15]: I think we should also explain to people what to do if they want to use the previous version of the gem (`solidus_braintree`). Is that stated anywhere at the moment?

File: lib/generators/solidus_braintree/update/update_generator.rb [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+      end
+
+      def run_migrations
+        run_migrations = options[:auto_run_migrations] || ['', 'y', 'Y'].include?(ask('Would you like to run the migrations now? [Y/n]')) # rubocop:disable Layout/LineLength
+        if run_migrations
+          run 'bin/rails db:migrate'
+        else
+          puts 'Skipping bin/rails db:migrate, don\'t forget to run it!' # rubocop:disable Rails/Output
+        end
+      end
```
Comment (kennyadsl) [inline lib/generators/solidus_braintree/update/update_generator.rb:20]: Nice, do you think we should double-check that people are not adding `solidus_braintree` and `solidus_paypal_braintree` together in their Gemfile? Not sure this installer is the right place, though.

**Review by kennyadsl — COMMENTED** [review]
Left some comments, but I can't see big blockers. Thanks!

File: app/assets/stylesheets/spree/backend/solidus_paypal_braintree.scss (hunk 2 above)
Comment (gsmendoza) [reply app/assets/stylesheets/spree/backend/solidus_paypal_braintree.scss:2]: Thanks @kennyadsl I'll try that!

File: README.md (hunk 7 above)
Comment (gsmendoza) [reply README.md:38]: @kennyadsl Hmmm... I'll try first adding a post-install message to the message linking to the README update section. Will let you know how that works out!

Comment (kennyadsl) [reply README.md:38]: But users don't need to see them when installing the gem, but after running the update command, or am I missing something?

Comment (gsmendoza) [reply README.md:38]: I'm thinking that when they switch from `solidus_paypal_braintree` to `solidus_braintree`, the post-install message would appear. But I'm not sure. Will test it out.

File: README.md (hunk 9 above)
Comment (gsmendoza) [reply README.md:15]: Probably makes sense to add information about the previous version in the README. I'll see what I can do.

File: lib/generators/solidus_braintree/update/update_generator.rb (hunk 10 above)
Comment (gsmendoza) [reply lib/generators/solidus_braintree/update/update_generator.rb:20]: @kennyadsl I would consider this check to be a nice-to-have, since it seems like an edge case. I imagine this can only happen when the developer forgets to remove the `solidus_paypal_braintree` line from the Gemfile. Are there other more common cases that I'm missing?

I'll test it out, to see what the experience for the user.

File: db/migrate/20230109080950_rename_solidus_paypal_braintree_source_type.rb [hunk 11]
```diff
@@ -0,0 +1,16 @@
+class RenameSolidusPaypalBraintreeSourceType < ActiveRecord::Migration[6.1]
```
Comment (elia) [inline db/migrate/20230109080950_rename_solidus_paypal_braintree_source_type.rb:1]: Should we also add a migration for changing the table names? Or are we aiming at handling everything with the table_name_prefix configuration?

**Review by elia — COMMENTED** [review]
🙌 Awesome, not a fun one to tackle!

Left a few comments and questions, anything marked as `nit` is not blocking 👍

File: lib/generators/solidus_braintree/update/update_generator.rb (hunk 10 above)
Comment (kennyadsl) [reply lib/generators/solidus_braintree/update/update_generator.rb:20]: Ouch, this made me think to a (maybe) giant issue. What if someone is currently using both payment methods? Maybe one for credit cart payments and one of paypal payments? I don't think it's that so unrealistic (@spaghetticode do you know any store in that scenario? 🙈). 

Well if that's the case they are done, because they can't add two gems with the same name to the Gemfile. What's the path for them? Do we want to consider it?

File: db/migrate/20160830061749_create_solidus_paypal_braintree_sources.rb [hunk 12]
```diff
@@ -1,6 +1,6 @@
...
   def change
-    create_table :solidus_paypal_braintree_sources do |t|
+    create_table :solidus_braintree_sources do |t|
```
[the remark this answers is not in the document]
Comment (gsmendoza) [reply db/migrate/20160830061749_create_solidus_paypal_braintree_sources.rb:3]: @elia I think you're right :) Let's rename the tables by adding more migrations. I also think that with those migrations, we can simplify the PR by removing the `table_name_prefix` configuration.

@kennyadsl Your thoughts?

File: README.md (hunk 9 above)
Comment (gsmendoza) [reply README.md:15]: > Yes, it's not advised to upgrade solidus_braintree from v4 to v5, rather it's intended for new stores and migrations from solidus_paypal_braintree.

Added a TODO for that in the description.

File: lib/generators/solidus_braintree/update/update_generator.rb (hunk 10 above)
Comment (gsmendoza) [reply lib/generators/solidus_braintree/update/update_generator.rb:20]: @kennyadsl The previous edge version of SolidusBraintree had a [warning](https://github.com/solidusio/solidus_braintree/tree/v1.x#deprecation-notice-warning-construction) up for two years that is already deprecated in favor of SolidusPaypalBraintree. Thus, if you are still using SolidusBraintree, you should be aware that there's a high risk in continuing to use it.

However, if you're using both SolidusBraintree and SolidusPaypalBraintree, I'm guessing the path would be 

1. Uninstall the old SolidusBraintree.
2. Replace SolidusPaypalBraintree with the new SolidusBraintree.
3. Either enable Paypal with SolidusBraintree or install SolidusPaypalCommercePlatform.

It seems that it will take a lot more effort to automate that migration path. Whether we want to implement it depends on how many people are still using the old SolidusBraintree.

Your thoughts?

File: lib/solidus_braintree/extension_configuration.rb [hunk 13]
```diff
@@ -7,6 +7,12 @@ class ExtensionConfiguration
     # Define here the settings for this extension, e.g.:
     #
     # attr_accessor :my_setting
+
+    attr_writer :table_name_prefix
+
+    def table_name_prefix
+      @table_name_prefix || 'solidus_braintree_'
+    end
```
[the remark this answers is not in the document]
Comment (gsmendoza) [reply lib/solidus_braintree/extension_configuration.rb:15]: I'm inclined to remove `table_name_prefix` in favor of new migrations renaming the tables. Do you think it's still worth it to keep the new configuration?

File: README.md [hunk 14]
```diff
@@ -22,6 +22,23 @@ bundle
 bundle exec rails g solidus_braintree:install
 ```
 
+## Upgrading to 2.0.0
```
[the remark this answers is not in the document]
Comment (gsmendoza) [reply README.md:25]: @elia I'm confused. Based on https://rubygems.org/gems/solidus_braintree and https://rubygems.org/gems/solidus_paypal_braintree, both versions of the gem are at v1.2.0. Am I missing something?

> If we want to let solidus_paypal_braintree users know there's such an upgrade we could release solidus_paypal_braintree v2 consisting of an empty gem that just depends on solidus_braintree v5 with a post-install message that explains the situation.

I think that's still covered by the current plan :) See https://github.com/solidusio/solidus_paypal_braintree/issues/333. We'll do it once we complete https://github.com/solidusio/solidus_braintree/issues/94.

File: db/migrate/20230109080950_rename_solidus_paypal_braintree_source_type.rb (hunk 11 above)
Comment (gsmendoza) [reply db/migrate/20230109080950_rename_solidus_paypal_braintree_source_type.rb:1]: As I mentioned in https://github.com/solidusio/solidus_braintree/pull/99#discussion_r1070197261, I'm inclined to remove `table_name_prefix` in favor of new migrations renaming the tables.

File: README.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (elia) [reply README.md:25]: 🤦‍♂️🤦‍♂️🤦‍♂️ Sorry, did the review on a Friday and probably looked at the wrong page (possibly solidus_stripe) of rubygems, please ignore all my comments about v4 and v5 ✂️

File: README.md [hunk 15]
```diff
@@ -22,6 +22,23 @@ bundle
...
+
+In v2.0.0, we've renamed the gem to SolidusPaypalBraintree to SolidusBraintree. To update your app:
+
+1. Change the gem in your Gemfile from 'solidus_paypal_braintree' to 'solidus_braintree'.
```
[the remark this answers is not in the document]
Comment (elia) [reply README.md:29]: Of course as long as the version is correct 🤦‍♂️ `5.0` → `2.0`

File: db/migrate/20230109080950_rename_solidus_paypal_braintree_source_type.rb (hunk 11 above)
Comment (elia) [reply db/migrate/20230109080950_rename_solidus_paypal_braintree_source_type.rb:1]: I think the migration should be enough, as long as it's crafted carefully, with stores that can have a huge number of records in mind (the `strong_migrations` gem can help in doing that).

File: lib/generators/solidus_braintree/update/update_generator.rb (hunk 10 above)
Comment (elia) [reply lib/generators/solidus_braintree/update/update_generator.rb:20]: @kennyadsl @gsmendoza I think this edge case should just be ignored, anyone in such a dire situation would probably just stay on the currently released versions and call it a day. 

And of course very likely that's not an actual use case.

File: lib/generators/solidus_braintree/update/update_generator.rb (hunk 6 above)
Comment (elia) [reply lib/generators/solidus_braintree/update/update_generator.rb:5]: Do we really need to use a different generator here? I'm thinking that everywhere else we used the install generator for both, relying on Thor's ability to propose diffs for conflicts and ignore identical files.

Thoughts on only using one?

Comment (gsmendoza) [reply lib/generators/solidus_braintree/update/update_generator.rb:5]: Okay, let me see if I can merge it with the InstallGenerator :+1:

File: lib/generators/solidus_braintree/update/update_generator.rb (hunk 10 above)
Comment (gsmendoza) [reply lib/generators/solidus_braintree/update/update_generator.rb:20]: @kennyadsl I intend to ignore for now the edge case of using both payment methods. Do you think we should still consider it?

File: db/migrate/20230109080950_rename_solidus_paypal_braintree_source_type.rb (hunk 11 above)
Comment (gsmendoza) [reply db/migrate/20230109080950_rename_solidus_paypal_braintree_source_type.rb:1]: @kennyadsl @elia Hmm... I haven't considered the potential effect of a table name to a large production database. `strong_migrations` says that renaming a table is potentially dangerous, so I'd rather leave it to the extension user to migrate their own database, rather than providing them a migration that can negatively impact them. 

I'll no longer continue with the table rename migration and will just keep the `table_name_prefix`.

File: lib/solidus_braintree/extension_configuration.rb (hunk 13 above)
[the remark this answers is not in the document]
Comment (gsmendoza) [reply lib/solidus_braintree/extension_configuration.rb:15]: Linking to https://github.com/solidusio/solidus_braintree/pull/99#discussion_r1071852119 - we'll keep the `table_name_prefix` and no longer continue with the table rename.

File: db/migrate/20160830061749_create_solidus_paypal_braintree_sources.rb (hunk 12 above)
[the remark this answers is not in the document]
Comment (kennyadsl) [reply db/migrate/20160830061749_create_solidus_paypal_braintree_sources.rb:3]: If we keep the configuration, are we able to let people upgrade and migrate data (in production) in two different moments? If the answer is no (as I think), I'd say that the configuration is useless. If yes, I'd keep it!

File: app/assets/stylesheets/spree/backend/solidus_paypal_braintree.scss (hunk 2 above)
Comment (gsmendoza) [reply app/assets/stylesheets/spree/backend/solidus_paypal_braintree.scss]: I updated the install generator to rename the asset files, so there is no more need for the deprecated asset files :+1:

File: app/assets/javascripts/spree/backend/solidus_paypal_braintree.js (hunk 3 above)
Comment (gsmendoza) [reply app/assets/javascripts/spree/backend/solidus_paypal_braintree.js]: The install generator now renames the references to the asset files, so there's no more need to keep the deprecated asset files.

File: db/migrate/20160830061749_create_solidus_paypal_braintree_sources.rb (hunk 12 above)
[the remark this answers is not in the document]
Comment (gsmendoza) [reply db/migrate/20160830061749_create_solidus_paypal_braintree_sources.rb:3]: @kennyadsl Linking to https://github.com/solidusio/solidus_braintree/pull/99#discussion_r1071852119 - we'll keep the `table_name_prefix` and no longer continue with the table rename.

File: solidus_braintree.gemspec [hunk 16]
```diff
@@ -39,4 +39,6 @@ Gem::Specification.new do |spec|
   spec.add_development_dependency 'solidus_dev_support', '~> 2.5'
   spec.add_development_dependency 'vcr'
   spec.add_development_dependency 'webmock'
+
+  spec.post_install_message = "If you're upgrading to v2.0.0, please see the README for upgrade instructions."
```
Comment (gsmendoza) [inline solidus_braintree.gemspec:43]: Once you rename `solidus_paypal_braintree` to `solidus_braintree` and run `bundle install`, you would see the post install message:

```
solidus_store 16:56:01 $ bundle install
Fetching https://github.com/solidusio/solidus_braintree.git
Fetching gem metadata from https://rubygems.org/.........
Resolving dependencies....
# ...
Using solidus_braintree 1.2.0 from https://github.com/solidusio/solidus_braintree.git (at gsmendoza/96-update-the-extension-namespace-to-solidus-braintree@2ef5bfe)
# ...
Bundle complete! 19 Gemfile dependencies, 131 gems now installed.
Use `bundle info [gemname]` to see where a bundled gem is installed.
Post-install message from solidus_braintree:
If you're upgrading to v2.0.0, please see the README for upgrade instructions.
```

File: lib/generators/solidus_braintree/install/install_generator.rb [hunk 17]
```diff
@@ -21,19 +21,17 @@ def add_stylesheets
...
       def add_migrations
-        run 'bin/rails railties:install:migrations FROM=solidus_braintree'
+        rake 'railties:install:migrations FROM=solidus_braintree'
       end
 
       def mount_engine
-        insert_into_file File.join('config', 'routes.rb'), after: "Rails.application.routes.draw do\n" do
-          "mount SolidusBraintree::Engine, at: '/solidus_braintree'\n"
-        end
+        route "mount SolidusBraintree::Engine, at: '/solidus_braintree'"
```
Comment (waiting-for-dev) [inline lib/generators/solidus_braintree/install/install_generator.rb:58]: Nit: Should the commit message say `Use rake and route` instead of `Use rake and mount`?

File: lib/solidus_braintree.rb [hunk 18]
```diff
@@ -8,6 +8,7 @@
 require 'solidus_braintree/extension_configuration'
 require 'solidus_braintree/version'
 require 'solidus_braintree/engine'
+require 'solidus_paypal_braintree'
```
Comment (waiting-for-dev) [inline lib/solidus_braintree.rb:11]: Have we considered not including this file by default? The idea is not to have the alias for users that don't need it. Would it make the upgrade path more difficult?

File: db/migrate/20230109080950_rename_solidus_paypal_braintree_source_type.rb [hunk 19]
```diff
@@ -0,0 +1,16 @@
...
+  def up
+    Spree::Payment.where(source_type: 'SolidusPaypalBraintree::Source')
+      .update_all("source_type = 'SolidusBraintree::Source'")
```
Comment (waiting-for-dev) [inline db/migrate/20230109080950_rename_solidus_paypal_braintree_source_type.rb:4]: Concerns that these `.update_all` queries could lock the table for too long? See https://github.com/ankane/strong_migrations#backfilling-data.

**Review by waiting-for-dev — COMMENTED** [review]
Thanks, @gsmendoza, left some questions.

File: lib/generators/solidus_braintree/install/install_generator.rb (hunk 17 above)
Comment (gsmendoza) [reply lib/generators/solidus_braintree/install/install_generator.rb:58]: Yep, you're right. I'll update the commit message :+1:

File: lib/solidus_braintree.rb (hunk 18 above)
Comment (gsmendoza) [reply lib/solidus_braintree.rb:11]: @waiting-for-dev I'll test if the user can manually add it after installing the extension's version update.

File: db/migrate/20230109080950_rename_solidus_paypal_braintree_source_type.rb (hunk 19 above)
Comment (gsmendoza) [reply db/migrate/20230109080950_rename_solidus_paypal_braintree_source_type.rb:4]: @waiting-for-dev Okay, I'll go ahead and comment them out with a warning saying that they can lock the tables affected. What do you think?

Comment (waiting-for-dev) [reply db/migrate/20230109080950_rename_solidus_paypal_braintree_source_type.rb:4]: Another option would be implementing them in the way `strong_migrations` recommends. I.e., no transaction, batching and throttling. Pinging @kennyadsl & @elia for more opinions.

Comment (kennyadsl) [reply db/migrate/20230109080950_rename_solidus_paypal_braintree_source_type.rb:4]: I can't see why not use the strong migration suggestion immediately. I was about to suggest this code:

```ruby
Spree::Payment.where(...).in_batches.update_all(...)
```

But their (similar) version seems to be even safer and considering more scenarios.

**Comment by gsmendoza** [comment]
@waiting-for-dev I have implemented "Post-review TODOs (2023-01-31)" in the description by rebasing the following commits:

* Add optional SolidusPaypalBraintree alias
* Use rake and route methods in InstallGenerator
* Update InstallGenerator to update to SolidusBraintree when applicable
* Add update instructions

I tested the  SolidusPaypalBraintree alias in https://github.com/gsmendoza/solidus_store/tree/rails-7-0-4--solidus-3-2-5--braintree--renamed--minimal--with-solidus-paypal-braintree. The user may need to add `require: 'solidus_paypal_braintree'` to their Gemfile first before installing the update (`rails g solidus_braintree:install`).

File: lib/generators/solidus_braintree/update/update_generator.rb (hunk 10 above)
Comment (gsmendoza) [reply lib/generators/solidus_braintree/update/update_generator.rb]: @kennyadsl I have chosen to skip testing the edge case where the user has both payment methods (old SolidusBraintree and SolidusPaypalBraintree), since based on this conversation thread, we think that it's highly unlikely. Let me know if you still have concerns about this.

**Review by waiting-for-dev — APPROVED** [review]
Awesome work, @gsmendoza 👏 👏

**Review by kennyadsl — APPROVED** [review]
👍 Thanks!

**Comment by gsmendoza** [comment]
Thank you @waiting-for-dev , @kennyadsl ! I confirmed in my manual test that the updated data migrations works as expected. I'm merging the PR now.
